pynitrokey/updates.py

```diff
@@ -38,12 +47,27 @@ def download_to_dir(self, d: str, overwrite: bool = False) -> str:
         if os.path.exists(path) and not overwrite:
             raise Exception(f"File {path} already exists and may not be overwritten")
         with open(path, "wb") as f:
-            self.download(f)
+            self.download(f, callback=callback)
         return path
 
-    def read(self) -> bytes:
-        with self._get() as response:
-            return response.content
+    def read(self, callback: Optional[ProgressCallback] = None) -> bytes:
+        result = bytes()
+        for chunk in self._get_chunks(callback=callback):
+            result += chunk
+        return result
+
+    def _get_chunks(
+        self, chunk_size: int = 1024, callback: Optional[ProgressCallback] = None
+    ) -> Generator[bytes, None, None]:
+        with self._get(stream=True) as response:
+            total = int(response.headers.get("content-length", 0))
+            if callback:
+                callback(0, total)
+
+            for chunk in response.iter_content(chunk_size=chunk_size):
+                if callback:
+                    callback(len(chunk), total)
+                yield chunk
```

Comment:
nit: this could have been potentially written simpler, given the `tqdm` example (linked below). I agree that current approach is more flexible though.
- https://pypi.org/project/tqdm/
```
import requests, os
from tqdm import tqdm

eg_link = "https://caspersci.uk.to/matryoshka.zip"
response = requests.get(eg_link, stream=True)
with tqdm.wrapattr(open(os.devnull, "wb"), "write",
                   miniters=1, desc=eg_link.split('/')[-1],
                   total=int(response.headers.get('content-length', 0))) as fout:
    for chunk in response.iter_content(chunk_size=4096):
        fout.write(chunk)
```

Reply:
As this code is in the general library part and not the CLI part of pynitrokey, I wanted to avoid using tqdm directly.